### PR TITLE
feat(batch): add support to SQS FIFO queues (SqsFifoPartialProcessor)

### LINKED ISSUE #1927: Feature request: Handle SQS FIFO queues with the Bach Processing utility
### Discussed in https://github.com/awslabs/aws-lambda-powertools-python/discussions/1913

<div type='discussions-op-text'>

<sup>Originally posted by **whoDoneItAgain** February  8, 2023</sup>
I've been reading over the documentation and I cant find an answer.

Does Powertools stop processing items after a failed and marks the failed plus all subsequent as failed in its response back to SQS?

Ex. r1,r2,r3,r4,r5 is sent. function fails on r3. ReportBatchItemFailures includes r3 as well as r4 and r5.

</div>

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #1927

### Changes

> Please provide a summary of what's being changed

This PR adds support for SQS FIFO queues on the Batch Processing utility.

From [docs](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#services-sqs-batchfailurereporting):

> If you're using this feature with a FIFO queue, your function should stop processing messages after the first failure and return all failed and unprocessed messages in `batchItemFailures`. This helps preserve the ordering of messages in your queue.

### User experience

> Please share what the user experience looks like before and after this change

Before this change, using the `BatchProcessor` with an SQS FIFO would result into undefined behaviour, as we were still continuing to process records after the first failure.

Using the new specialized class `SqsFifoPartialProcessor` makes the utility safe to use with SQS FIFO.

carbon (37)

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import sys
+from typing import List, Optional, Tuple, Type
+
+from aws_lambda_powertools.utilities.batch import BatchProcessor, EventType
+from aws_lambda_powertools.utilities.parser.models import SqsRecordModel
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: this is gonna fail if it doesn't have pydantic. Let's move this entire logic to a `types.py` as it's spans beyond `base.py` now

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+class SQSFifoCircuitBreakerError(Exception):
+    """
+    Signals a record not processed due to the SQS FIFO processing being interrupted
+    """
+
+    pass
+
+
+class SQSFifoPartialProcessor(BatchProcessor):
+    """Specialized BatchProcessor subclass that handles FIFO SQS batch records.
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: Shorten that. Focus on explain the FIFO element (less on subclassing, that's internal)

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    As soon as the processing of the first record fails, the remaining records
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: We can do better here - let's shorten it and explain in layman terms what this all means.

Always think of a non-developer persona reading this for the first time to understand whether they should use it or not. Once our new API reference is live, this will be the headline.

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    ```python
+    import json
+
+    from aws_lambda_powertools import Logger, Tracer
+    from aws_lambda_powertools.utilities.batch import SQSFifoPartialProcessor, EventType, batch_processor
+    from aws_lambda_powertools.utilities.data_classes.sqs_event import SQSRecord
+    from aws_lambda_powertools.utilities.typing import LambdaContext
+
+
+    processor = SQSFifoPartialProcessor()
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: will need change once class is renamed

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        super().__init__(EventType.SQS, model)
+
+    def process(self) -> List[Tuple]:
+        result: List[Tuple] = []
+
+        for i, record in enumerate(self.records):
+            """
+            If we have failed messages, it means that the last message failed.
+            We then short circuit the process, failing the remaining messages
+            """
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: that's interesting take, TIL! I wonder if these should be `#` or overwrite the docstring for `def process` since it differs from the typical `BatchProcessor::process()`

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+        return result
+
+    def _short_circuit_processing(self, first_failure_index: int, result: List[Tuple]) -> List[Tuple]:
+        remaining_records = self.records[first_failure_index:]
+        for remaining_record in remaining_records:
+            data = self._to_batch_type(record=remaining_record, event_type=self.event_type, model=self.model)
+            result.append(
+                self.failure_handler(
+                    record=data, exception=(type(self.circuitBreakerError), self.circuitBreakerError, None)
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: Since the tuple creation is static (type(self.circuitBreakerError), self.circuitBreakerError, None), create it earlier where `_short_circuit_processing` can access it - same way you micro-optimized the `circuitBreakerError` as a class attr.

Creating these tuples in a large data pipeline will use unnecessary memory. It'll also have a nice side-effect of making it easier to read code:

```python
result.append(self.failure_handler(record=data, exception=self.circuit_break_exc)
```

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            """
+            If we have failed messages, it means that the last message failed.
+            We then short circuit the process, failing the remaining messages
+            """
+            if self.fail_messages:
+                return self._short_circuit_processing(i, result)
+
+            """
+            Otherwise, process the message normally
+            """
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]:
```suggestion
	   # Otherwise, process the message normally
```

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            result.append(self._process_record(record))
+
+        return result
+
+    def _short_circuit_processing(self, first_failure_index: int, result: List[Tuple]) -> List[Tuple]:
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py:81]: Q: do we need a docstring? It's clear to me, is it for @leandrodamascena ?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
This is great!! Made some initial comments and spotted two issues. After lunch, I'll look over docs and tests.

Thank you @rubenfonseca <3

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py (hunk 4 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: @heitorlessa what would you suggest the new class name would be? I believe I missed that on the review

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py (hunk 8 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py:81]: Added a docstring to make it clearer

File: aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py (hunk 4 above)
Comment (heitorlessa) [reply aws_lambda_powertools/utilities/batch/sqs_fifo_partial_processor.py]: ahh it got lost somehow - I suggested either `SqsFifoPartialProcessor` or `SqsFifoProcessor`. `SQS` isn't aligned with other parts of the code.

File: aws_lambda_powertools/utilities/batch/types.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        KinesisDataStreamRecord as KinesisDataStreamRecordModel,
+    )
+    from aws_lambda_powertools.utilities.parser.models import SqsRecordModel
+
+    BatchTypeModels = Optional[
+        Union[Type[SqsRecordModel], Type[DynamoDBStreamRecordModel], Type[KinesisDataStreamRecordModel]]
+    ]
+    BatchSqsTypeModel = Optional[Type[SqsRecordModel]]
+else:
+    BatchTypeModels = Optional  # type: ignore
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/batch/types.py]: did you mean `BatchTypeModels = "BatchTypeModels"` like we had previously?

File: docs/utilities/batch.md [hunk 10]
```diff
@@ -347,6 +347,24 @@ Processing batches from SQS works in four stages:
     }
     ```
 
+#### FIFO queues
+
+If you're using this feature with a FIFO queue, you should use the `SQSFifoPartialProcessor` class instead. We will
```
Comment (heitorlessa) [inline docs/utilities/batch.md]:
```suggestion
When using [SQS FIFO queues](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html){target="_blank"}, use the `SQSFifoPartialProcessor` class instead. We will
```

File: tests/functional/test_utilities_batch.py [hunk 11]
```diff
@@ -40,7 +42,7 @@
 def sqs_event_factory() -> Callable:
     def factory(body: str):
         return {
-            "messageId": "059f36b4-87a3-44ab-83d2-661975830a7d",
+            "messageId": str(uuid.uuid4()),
```
Comment (heitorlessa) [inline tests/functional/test_utilities_batch.py]:
```suggestion
            "messageId": f"{uuid.uuid4()}",
```

File: tests/functional/test_utilities_batch.py [hunk 12]
```diff
@@ -117,6 +119,17 @@ def handler(record):
...
 
+@pytest.fixture(scope="module")
+def sqs_fifo_record_handler() -> Callable:
+    def handler(record):
+        body = record["body"]
+        if "fail" in body:
+            raise Exception("Failed to process record.")
+        return body
+
+    return handler
```
Comment (heitorlessa) [inline tests/functional/test_utilities_batch.py]: naive question, do we need it given it's the exact same logic as `record_handler` fixture?

File: tests/functional/test_utilities_batch.py [hunk 13]
```diff
@@ -654,6 +667,48 @@ def lambda_handler(event, context):
...
+
+    # THEN
+    assert result["batchItemFailures"] == []
+
+
+def test_sqs_fifo_batch_processor_middleware_with_failure(sqs_event_factory, sqs_fifo_record_handler):
+    # GIVEN
+    first_record = SQSRecord(sqs_event_factory("success"))
+    second_record = SQSRecord(sqs_event_factory("fail"))
+    # this would normally suceed, but since it's a FIFO queue, it will be marked as failure
```
Comment (heitorlessa) [inline tests/functional/test_utilities_batch.py]: loved this comment. thank you for your great attention to detail @rubenfonseca

File: aws_lambda_powertools/utilities/batch/types.py (hunk 9 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/utilities/batch/types.py]: Yes that makes much more sense, thank you!

File: docs/utilities/batch.md (hunk 10 above)
Comment (rubenfonseca) [reply docs/utilities/batch.md]: Much better, thanks!

File: tests/functional/test_utilities_batch.py (hunk 12 above)
Comment (rubenfonseca) [reply tests/functional/test_utilities_batch.py]: Absolutely right. I created it in the beginning because I thought it would be different, but it's not :) Changing it

File: docs/utilities/batch.md [hunk 14]
```diff
@@ -349,8 +349,8 @@ Processing batches from SQS works in four stages:
 
 #### FIFO queues
 
-If you're using this feature with a FIFO queue, you should use the `SQSFifoPartialProcessor` class instead. We will
-stop processing messages after the first failure, and return all failed and unprocessed messages in `batchItemFailures`.
+When using [SQS FIFO queues](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html){target="_blank"},
+We will stop processing messages after the first failure, and return all failed and unprocessed messages in `batchItemFailures`.
```
Comment (heitorlessa) [inline docs/utilities/batch.md]:
```suggestion
When using [SQS FIFO queues](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html){target="_blank"}, we will stop processing messages after the first failure, and return all failed and unprocessed messages in `batchItemFailures`.
```

**Review by heitorlessa — APPROVED** [review]
